File: tests/native/src/js/arrays.rs
use neon::vm::{Call, JsResult};
use neon::mem::Handle;
use neon::js::{JsNumber, JsString, JsArray, Object};

pub fn return_js_array(call: Call) -> JsResult<JsArray> {
    Ok(JsArray::new(call.scope, 0))
}

pub fn return_js_array_with_number(call: Call) -> JsResult<JsArray> {
    let scope = call.scope;
    let array: Handle<JsArray> = JsArray::new(scope, 1);
    array.set(0, JsNumber::new(scope, 9000))?;
    Ok(array)
}

pub fn return_js_array_with_string(call: Call) -> JsResult<JsArray> {
    let scope = call.scope;
    let array: Handle<JsArray> = JsArray::new(scope, 1);
    array.set(0, JsString::new(scope, "hello node").unwrap())?;
    Ok(array)
}

File: test/napi/lib/arrays.js
var addon = require('../native');
var assert = require('chai').assert;

describe('JsArray', function() {
  it('return a JsArray built in Rust', function () {
    assert.deepEqual([], addon.return_js_array());
  });

  it('return a JsArray with a number at index 0', function () {
    assert.deepEqual([9000], addon.return_js_array_with_number());
  });

  it('return a JsArray with an string at index 0', function () {
    assert.deepEqual(["hello node"], addon.return_js_array_with_string());
  });

  it('can read from a JsArray', function () {
    assert.strictEqual(addon.read_js_array([1234]), 1234);
  });

  it('returns undefined when accessing outside JsArray bounds', function () {
    assert.strictEqual(addon.read_js_array([]), undefined);
  });
});
